[tensorexpr] Allocate intermediate buffers at compile time if specified

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64227
* #64226



Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

huiguoo committed:
[tensorexpr] Allocate intermediate buffers at compile time

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -241,6 +241,9 @@ class TORCH_API TensorExprKernel {
   }
 
+  // Allocate memory for intermediate buffers at compile time.
+  void preAllocIntermediateBufs(const std::unordered_set<BufPtr> intem_bufs);
```
```suggestion
  void preAllocIntermediateBufs(const std::unordered_set<BufPtr>& intem_bufs);
```

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -241,6 +241,9 @@ class TORCH_API TensorExprKernel {
     return custom_lowerings_;
   }
 
+  // Allocate memory for intermediate buffers at compile time.
```
Could you please add an explanation how it works? It might be non-obvious that preallocated bufs end up among constants (which is a totally right thing to do here, just not obvious).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3036,6 +3038,39 @@ void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
+        is_static = false;
+        break;
+      }
+      size = size * immediateAs<int>(d);
```
With @bertmaher's #64077 it might be better to use `immLike` function here so that it works for 64-bit ints as well (dimensions might be 64-bit).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3036,6 +3038,39 @@ void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
+
+void TensorExprKernel::preAllocIntermediateBufs(
+    const std::unordered_set<BufPtr> intem_bufs) {
+  void* bp;
```
Nit: this can be declared inside the loop.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2726,6 +2726,8 @@ StmtPtr TensorExprKernel::transformLoops(BackendType backendType, StmtPtr st) {
     }
   }
 
+  auto intem_bufs = l.getIntermediateBufs();
```
Nit: should it be "inte_**r**_m" instead of "intem"?

ZolotukhinM commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -85,6 +85,41 @@ TEST_F(Kernel, InliningIntermediates) {
+  oss << *stmt;
+
+  // Check the IR we produced
+  torch::jit::testing::FileCheck().check_not("Alloc")->run(oss.str());
+  torch::jit::testing::FileCheck().check_not("Free")->run(oss.str());
```
Might be worth also checking that it ended up among constants (we can inspect constants with `getConstantDescriptors`).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3036,6 +3038,39 @@ void TensorExprKernel::bindConstant(const torch::jit::Value* v) {
+void TensorExprKernel::preAllocIntermediateBufs(
+    const std::unordered_set<BufPtr> intem_bufs) {
+  void* bp;
+  for (auto buf : intem_bufs) {
```
Iterating over unordered_set yields non-deterministic results - i.e. the order in which the buffers get into constants_ is different each time we run. While this is correct each time, it's often quite annoying to deal with that when, e.g., we need to debug something, because execution might be different from run to run. It is usually recommended to put all elements into a vector and sort it before iterating over them to avoid this.

ZolotukhinM reviewed (commented):
Awesome! I have some comments below, and also a question on the previous PR, please take a look!

bertmaher reviewed (changes requested):
I don't think this is thread safe.  If multiple threads running a kernel at once, aren't they all going to step on the same intermediate buffer?

Can you describe the use case exactly?

huiguoo commented:
Sorry for the confusion @bertmaher. This PR is unrelated to the parallel use case discussed in the chat. It only switches the allocation of intermediate buffers from runtime to compile time. Before it, all intermediate buffers are global and shared by multiple threads and this remains the same after the change. 

The parallel use case is from the 5x5 convs (#63168) : I attempted to parallelize the outer loop while scheduling but there is WAW dependency on an intermediate buffer in the loop. So I tried to make that buffer thread-private by inserting `Alloc` inside the loop. But the intermediate buffer is still global so the results are incorrect, and as you mentioned it causes memory leaks. @ZolotukhinM pointed me to the parallel codegen https://github.com/pytorch/pytorch/blob/968d7ee46a66316557a04a64333d2810b544ed9b/torch/csrc/jit/tensorexpr/llvm_codegen.cpp#L1232 and suggested to add 'bufToVal' mapping but I haven't figured it out yet.

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

bertmaher commented:
> It only switches the allocation of intermediate buffers from runtime to compile time. Before it, all intermediate buffers are global and shared by multiple threads

The multithreaded case I'm worried about is the case where we have multiple concurrent executions of the kernel happening, not where we've done intra-kernel threading with `set_parallel`.  If we allocate just one buffer at compile time, burn the address into the kernel, and then try to execute that kernel from multiple threads, we're going to have a bad time.  I don't think it's possible to do allocation at compile time like this, unless we can guarantee externally that only one instance of this kernel will ever be in flight.

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

huiguoo commented:
Ah, that's a good point! @navahgar mentions static runtime is executing compiled NNC kernels concurrently in multiple threads but those are single-op kernels (and fused elementwise op kernels in the future) without intermediate bufs. How does mobile team use NNC @ZolotukhinM ?

ZolotukhinM commented:
Right, that's a very good point. AFAIU, mobile team is not interested in multi-threading for this case yet. I think adding this optimization as an optional step that only will be used in AOT pipeline is a good idea!

navahgar commented:
What does it mean to allocate memory at compile time for AOT compilation flow? Are you referring to statically allocating the memory for buffers? How will you support symbolic shapes in that case?

ZolotukhinM commented:
> What does it mean to allocate memory at compile time for AOT compilation flow?

The AOT pipeline doesn't rely on us actually allocating intermediate buffers, but they rely on us computing how much memory needs to be allocated. When we do this optimization, we record this info (along with the pointer to allocated memory) to constants_ and this is then put into memory-plan that mobile runtime uses when setting up a runtime for NNC-compiled model. This way if the model is run many times, allocations will only happen once in the very beginning when we're firing up the model.

https://github.com/pytorch/pytorch/blob/3cd0a4ac153ea8404f0e960ddacf00098689e600/torch/csrc/jit/mobile/nnc/context.cpp#L79-L87


If shapes are not statically known, preallocation will not happen. For such buffers the memory will be allocated dynamically, like it's supposed to.

navahgar commented:
Aah okay. So, we don't actually allocate memory at compile time. We just record this info so that the mobile runtime allocates memory on startup. Thanks for clarifying. I got confused by "allocation at compile-time".

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time if specified"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

huiguoo commented:
@ZolotukhinM I added a default parameter in TEK to let users specify whether they want intermediate buffers to be allocated at compile time. This (a default parameter) seems to be the easiest way to add compile-time allocation: once `compile()` is done, It becomes very hard to update the data in TEK. 

If the users prefer to manage the memory in their own way, we can replace this bool parameter to a function parameter in the future so the users can specify their own buffer allocation function.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -3151,10 +3198,14 @@ void TensorExprKernel::compile() {
     : graph_(subgraph),
       code_(subgraph, ""),
       custom_lowerings_(std::move(custom_lowerings)) {
+  if (pre_alloc) {
+    pre_alloc_ = true;
+  }
```
Why not just
```
pre_alloc_(pre_alloc)
```
in the constructors section?

ZolotukhinM reviewed (approved):
Looks good!

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time if specified"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time if specified"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).

huiguoo committed:
Update on "[tensorexpr] Allocate intermediate buffers at compile time if specified"




Differential Revision: [D30652220](https://our.internmc.facebook.com/intern/diff/D30652220)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652220).
